osu.Game/Rulesets/Difficulty/Skills/PreStrainSkill.cs

```diff
@@ -0,0 +1,39 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Collections.Generic;
+using osu.Game.Rulesets.Difficulty.Preprocessing;
+using osu.Game.Rulesets.Mods;
+
+namespace osu.Game.Rulesets.Difficulty.Skills
+{
+    public abstract class PreStrainSkill : StrainSkill
+    {
+        protected abstract double StrainDecayBase { get; }
+
+        protected abstract double SkillMultiplier { get; }
+
+        protected PreStrainSkill(Mod[] mods)
+            : base(mods)
+        {
+        }
+
+        public List<double> GetAllStrainPeaks() => StrainPeaks;
+
+        protected override void Process(int index, DifficultyHitObject current)
+        {
+            CurrentSectionPeak *= StrainDecayBase;
+            CurrentSectionPeak += StrainValueAt(index, current);
+            SaveCurrentPeak();
+        }
+
+        protected override double CalculateInitialStrain(double time) => 0;
+
+        public void ProcessPre(int index, DifficultyHitObject current)
+        {
+            ProcessInternal(index, current);
+        }
```

Comment:
i'm really confused by this. `ProcessPre()` calls `ProcessInternal()` which calls `Process()`. i do not understand why this exists when you may as well call `ProcessInternal()` directly?

also this index thing doesn't need to (and arguably shouldn't) exist. if skills need to be storing strain values from previous objects, you don't need to do any of this, just keep an internal list/limited size queue of the past values in the skill itself.

Reply:
do you want to visibility ProcessInternal() from protected to public?

if then why do they set the visibility to internal? no reason to hide it now

Reply:
if the only point of this added method is just to expose `ProcessInternal` as public, then yes, might as well

Reply:
yep

Reply:
but we can't get difficultyvalue() if it doesnt store all values

Reply:
Ideal solution:
1. Keep the array storing all values
2. The overloading changes to StrainPeaks[StrainPeaks.Count - 1 - i]
2-1. Deleting all index parameter

I will do it until tomorrow.